# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Import your libraries
import pandas as pd
import numpy as np
import sklearn


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. You can find the data in Ironhack's database (db: weather, table: austin_weather). 

In [5]:
# Your code here
austin=pd.read_csv('./austin_weather.csv')
austin

Date  TempHighF  TempAvgF  \
0     2013-12-21,74,60,45,67,49,43,93,75,57,29.86,29...        NaN       NaN   
1                                            2013-12-22       56.0      48.0   
2                                            2013-12-23       58.0      45.0   
3                                            2013-12-24       61.0      46.0   
4                                            2013-12-25       58.0      50.0   
...                                                 ...        ...       ...   
1314                                         2017-07-27      103.0      89.0   
1315                                         2017-07-28      105.0      91.0   
1316                                         2017-07-29      107.0      92.0   
1317                                         2017-07-30      106.0      93.0   
1318                                         2017-07-31       99.0      88.0   

      TempLowF DewPointHighF DewPointAvgF DewPointLowF HumidityHighPercent  \
0          NaN           NaN          NaN          NaN                 NaN   
1         39.0            43           36           28                  93   
2         32.0            31           27           23                  76   
3         31.0            36           28           21                  89   
4         41.0            44           40           36                  86   
...        ...           ...          ...          ...                 ...   
1314      75.0            71           67           61                  82   
1315      76.0            71           64           55                  87   
1316      77.0            72           64           55                  82   
1317      79.0            70           68           63                  69   
1318      77.0            66           61           54                  64   

     HumidityAvgPercent HumidityLowPercent  ... SeaLevelPressureAvgInches  \
0                   NaN                NaN  ...                       NaN   
1                    68                 43  ...                     30.13   
2                    52                 27  ...                     30.49   
3                    56                 22  ...                     30.45   
4                    71                 56  ...                     30.33   
...                 ...                ...  ...                       ...   
1314                 54                 25  ...                     29.97   
1315                 54                 20  ...                      29.9   
1316                 51                 19  ...                     29.86   
1317                 48                 27  ...                     29.91   
1318                 43                 22  ...                     29.97   

     SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                          NaN                 NaN                NaN   
1                        29.87                  10                 10   
2                        30.41                  10                 10   
3                         30.3                  10                 10   
4                        30.27                  10                 10   
...                        ...                 ...                ...   
1314                     29.88                  10                 10   
1315                     29.81                  10                 10   
1316                     29.79                  10                 10   
1317                     29.87                  10                 10   
1318                     29.91                  10                 10   

     VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                   NaN         NaN        NaN         NaN   
1                     5          16          6          25   
2                    10           8          3          12   
3                     7          12          4          20   
4                     7          10      

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [6]:
# Your code here
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Date                        1319 non-null   object 
 1   TempHighF                   1133 non-null   float64
 2   TempAvgF                    1133 non-null   float64
 3   TempLowF                    1133 non-null   float64
 4   DewPointHighF               1133 non-null   object 
 5   DewPointAvgF                1133 non-null   object 
 6   DewPointLowF                1133 non-null   object 
 7   HumidityHighPercent         1133 non-null   object 
 8   HumidityAvgPercent          1133 non-null   object 
 9   HumidityLowPercent          1133 non-null   object 
 10  SeaLevelPressureHighInches  1133 non-null   object 
 11  SeaLevelPressureAvgInches   1133 non-null   object 
 12  SeaLevelPressureLowInches   1133 non-null   object 
 13  VisibilityHighMiles         1133 

In [7]:
# Your code here
austin.isnull().sum()

Date                            0
TempHighF                     186
TempAvgF                      186
TempLowF                      186
DewPointHighF                 186
DewPointAvgF                  186
DewPointLowF                  186
HumidityHighPercent           186
HumidityAvgPercent            186
HumidityLowPercent            186
SeaLevelPressureHighInches    186
SeaLevelPressureAvgInches     186
SeaLevelPressureLowInches     186
VisibilityHighMiles           186
VisibilityAvgMiles            186
VisibilityLowMiles            186
WindHighMPH                   186
WindAvgMPH                    186
WindGustMPH                   186
PrecipitationSumInches        186
Events                        186
dtype: int64

In [8]:
# Your code here
austin.describe()

TempHighF     TempAvgF     TempLowF
count  1133.000000  1133.000000  1133.000000
mean     80.857899    70.379523    59.367167
std      15.083282    14.395438    14.533538
min      32.000000    29.000000    19.000000
25%      72.000000    61.000000    48.000000
50%      83.000000    73.000000    62.000000
75%      93.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [9]:
austin.head()

Date  TempHighF  TempAvgF  \
0  2013-12-21,74,60,45,67,49,43,93,75,57,29.86,29...        NaN       NaN   
1                                         2013-12-22       56.0      48.0   
2                                         2013-12-23       58.0      45.0   
3                                         2013-12-24       61.0      46.0   
4                                         2013-12-25       58.0      50.0   

   TempLowF DewPointHighF DewPointAvgF DewPointLowF HumidityHighPercent  \
0       NaN           NaN          NaN          NaN                 NaN   
1      39.0            43           36           28                  93   
2      32.0            31           27           23                  76   
3      31.0            36           28           21                  89   
4      41.0            44           40           36                  86   

  HumidityAvgPercent HumidityLowPercent  ... SeaLevelPressureAvgInches  \
0                NaN                NaN  ...                       NaN   
1                 68                 43  ...                     30.13   
2                 52                 27  ...                     30.49   
3                 56                 22  ...                     30.45   
4                 71                 56  ...                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                       NaN                 NaN                NaN   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                NaN         NaN        NaN         NaN   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches Events  
0                    NaN    NaN  
1                      0         
2                      0         
3                      0         
4                      T         

[5 rows x 21 columns]

In [10]:
austin['Events'].unique()

array([nan, ' ', 'Rain', 'Fog', 'Thunderstorm'], dtype=object)

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [11]:
# Your code here
austin['DewPointAvgF'].unique()

array([nan, '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '17', '14', '13', '54', '59', '15', '24', '29', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '65', '68', '69', '-',
       '72', '71', '73', '74', '49', '21', '52', '12', '44', '75', '76',
       '18'], dtype=object)

In [ ]:
# Your observation here


The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [12]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [13]:
# Your code here
austin[wrong_type_columns] = austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce')

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [14]:
# Your code here
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Date                        1319 non-null   object 
 1   TempHighF                   1133 non-null   float64
 2   TempAvgF                    1133 non-null   float64
 3   TempLowF                    1133 non-null   float64
 4   DewPointHighF               1126 non-null   float64
 5   DewPointAvgF                1126 non-null   float64
 6   DewPointLowF                1126 non-null   float64
 7   HumidityHighPercent         1131 non-null   float64
 8   HumidityAvgPercent          1131 non-null   float64
 9   HumidityLowPercent          1131 non-null   float64
 10  SeaLevelPressureHighInches  1130 non-null   float64
 11  SeaLevelPressureAvgInches   1130 non-null   float64
 12  SeaLevelPressureLowInches   1130 non-null   float64
 13  VisibilityHighMiles         1121 

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [15]:
# Your code here
missing_values=austin[austin.isnull().any(axis=1)==True]
missing_values

Date  TempHighF  TempAvgF  \
0     2013-12-21,74,60,45,67,49,43,93,75,57,29.86,29...        NaN       NaN   
4                                            2013-12-25       58.0      50.0   
6                                            2013-12-27       60.0      53.0   
7                                            2013-12-28       62.0      51.0   
33    2014-01-23,56,42,27,42,29,20,78,64,50,30.73,30...        NaN       NaN   
...                                                 ...        ...       ...   
1304  2017-07-17,98,88,77,75,71,66,88,63,38,29.94,29...        NaN       NaN   
1308                                         2017-07-21      104.0      91.0   
1310  2017-07-23,103,90,77,74,71,66,85,58,31,29.93,2...        NaN       NaN   
1311  2017-07-24,102,89,76,75,71,56,91,60,29,30.01,2...        NaN       NaN   
1313                                         2017-07-26      103.0      91.0   

      TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
0          NaN            NaN           NaN           NaN   
4         41.0           44.0          40.0          36.0   
6         45.0           41.0          39.0          37.0   
7         40.0           43.0          39.0          33.0   
33         NaN            NaN           NaN           NaN   
...        ...            ...           ...           ...   
1304       NaN            NaN           NaN           NaN   
1308      77.0           74.0          68.0          61.0   
1310       NaN            NaN           NaN           NaN   
1311       NaN            NaN           NaN           NaN   
1313      78.0           75.0          70.0          61.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  ...  \
0                     NaN                 NaN                 NaN  ...   
4                    86.0                71.0                56.0  ...   
6                    83.0                65.0                47.0  ...   
7                    92.0                64.0                36.0  ...   
33                    NaN                 NaN                 NaN  ...   
...                   ...                 ...                 ...  ...   
1304                  NaN                 NaN                 NaN  ...   
1308                 85.0                56.0                27.0  ...   
1310                  NaN                 NaN                 NaN  ...   
1311                  NaN                 NaN                 NaN  ...   
1313                 91.0                58.0                25.0  ...   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                           NaN                        NaN   
4                         30.33                      30.27   
6                         30.39                      30.34   
7                         30.17                      30.04   
33                          NaN                        NaN   
...                         ...                        ...   
1304                        NaN                        NaN   
1308                      29.95                      29.86   
1310                        NaN                        NaN   
1311                        NaN                        NaN   
1313                      30.03                      29.93   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                     NaN                 NaN                 NaN   
4                    10.0                10.0                 7.0   
6                    10.0                 9.0                 7.0   
7                    10.0                10.0                 7.0   
33                    NaN                 NaN                 NaN   
...                   ...                 ...                 ...   
1304                  NaN                 NaN                 NaN   
1308                 10.0                10.0                10.0   
1310                  NaN                 NaN                 NaN   
1311                  NaN        

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [16]:
# Your code here
len(austin.index)

1319

In [18]:
missing_values.count().max()

310

In [19]:
austin.isnull().sum()

Date                            0
TempHighF                     186
TempAvgF                      186
TempLowF                      186
DewPointHighF                 193
DewPointAvgF                  193
DewPointLowF                  193
HumidityHighPercent           188
HumidityAvgPercent            188
HumidityLowPercent            188
SeaLevelPressureHighInches    189
SeaLevelPressureAvgInches     189
SeaLevelPressureLowInches     189
VisibilityHighMiles           198
VisibilityAvgMiles            198
VisibilityLowMiles            198
WindHighMPH                   188
WindAvgMPH                    188
WindGustMPH                   190
PrecipitationSumInches        298
Events                        186
dtype: int64

#### Calculate the ratio of missing rows to total rows

In [20]:
# Your code here
percent_missing = missing_values.isnull().count().max()*100/austin.count().max()
percent_missing

23.502653525398028

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [21]:
# Your code here
austin.isna().sum()

Date                            0
TempHighF                     186
TempAvgF                      186
TempLowF                      186
DewPointHighF                 193
DewPointAvgF                  193
DewPointLowF                  193
HumidityHighPercent           188
HumidityAvgPercent            188
HumidityLowPercent            188
SeaLevelPressureHighInches    189
SeaLevelPressureAvgInches     189
SeaLevelPressureLowInches     189
VisibilityHighMiles           198
VisibilityAvgMiles            198
VisibilityLowMiles            198
WindHighMPH                   188
WindAvgMPH                    188
WindGustMPH                   190
PrecipitationSumInches        298
Events                        186
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [22]:
# Your code here
austin['PrecipitationSumInches'].isna().sum()*100/austin.count().max()

22.592873388931007

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [23]:
# Your code here 

austin.drop('PrecipitationSumInches',axis=1, inplace=True)
# Print `austin` to confirm the column is indeed removed

austin

Date  TempHighF  TempAvgF  \
0     2013-12-21,74,60,45,67,49,43,93,75,57,29.86,29...        NaN       NaN   
1                                            2013-12-22       56.0      48.0   
2                                            2013-12-23       58.0      45.0   
3                                            2013-12-24       61.0      46.0   
4                                            2013-12-25       58.0      50.0   
...                                                 ...        ...       ...   
1314                                         2017-07-27      103.0      89.0   
1315                                         2017-07-28      105.0      91.0   
1316                                         2017-07-29      107.0      92.0   
1317                                         2017-07-30      106.0      93.0   
1318                                         2017-07-31       99.0      88.0   

      TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
0          NaN            NaN           NaN           NaN   
1         39.0           43.0          36.0          28.0   
2         32.0           31.0          27.0          23.0   
3         31.0           36.0          28.0          21.0   
4         41.0           44.0          40.0          36.0   
...        ...            ...           ...           ...   
1314      75.0           71.0          67.0          61.0   
1315      76.0           71.0          64.0          55.0   
1316      77.0           72.0          64.0          55.0   
1317      79.0           70.0          68.0          63.0   
1318      77.0           66.0          61.0          54.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0                     NaN                 NaN                 NaN   
1                    93.0                68.0                43.0   
2                    76.0                52.0                27.0   
3                    89.0                56.0                22.0   
4                    86.0                71.0                56.0   
...                   ...                 ...                 ...   
1314                 82.0                54.0                25.0   
1315                 87.0                54.0                20.0   
1316                 82.0                51.0                19.0   
1317                 69.0                48.0                27.0   
1318                 64.0                43.0                22.0   

      SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                            NaN                        NaN   
1                          30.41                      30.13   
2                          30.56                      30.49   
3                          30.56                      30.45   
4                          30.41                      30.33   
...                          ...                        ...   
1314                       30.04                      29.97   
1315                       29.97                      29.90   
1316                       29.91                      29.86   
1317                       29.96                      29.91   
1318                       30.04                      29.97   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                           NaN                  NaN                 NaN   
1                         29.87                 10.0                10.0   
2                         30.41                 10.0                10.0   
3                         30.30                 10.0                10.0   
4                         30.27                 10.0                10.0   
...                         ...                  ...                 ...   
1314                      29.88                 10.0                10.0   
1315                      29.81                 10.0                10.0   
1316                      29.79                 10.0                10.0   
1317       

In [26]:
for col in austin.columns:
    print(col)

Date
TempHighF
TempAvgF
TempLowF
DewPointHighF
DewPointAvgF
DewPointLowF
HumidityHighPercent
HumidityAvgPercent
HumidityLowPercent
SeaLevelPressureHighInches
SeaLevelPressureAvgInches
SeaLevelPressureLowInches
VisibilityHighMiles
VisibilityAvgMiles
VisibilityLowMiles
WindHighMPH
WindAvgMPH
WindGustMPH
Events


#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [27]:
austin_fixed=austin.interpolate(inplace=False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [28]:
# Your code here
austin_fixed_missing=austin_fixed[austin_fixed.isnull().any(axis=1)==True]
print(austin_fixed_missing.count())
austin_missing=austin[austin.isnull().any(axis=1)==True]
print(austin_missing.count())

Date                          186
TempHighF                     185
TempAvgF                      185
TempLowF                      185
DewPointHighF                 185
DewPointAvgF                  185
DewPointLowF                  185
HumidityHighPercent           185
HumidityAvgPercent            185
HumidityLowPercent            185
SeaLevelPressureHighInches    185
SeaLevelPressureAvgInches     185
SeaLevelPressureLowInches     185
VisibilityHighMiles           185
VisibilityAvgMiles            185
VisibilityLowMiles            185
WindHighMPH                   185
WindAvgMPH                    185
WindGustMPH                   185
Events                          0
dtype: int64
Date                          200
TempHighF                      14
TempAvgF                       14
TempLowF                       14
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent            12
HumidityAvgPercent             12
H

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [29]:
# Your code here:
austin['Events'].value_counts()

                903
Rain            192
Fog              21
Thunderstorm     17
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [ ]:
# Your answer:


#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [30]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for i in event_list:
    austin_fixed[i]=0

# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  \
0     2013-12-21,74,60,45,67,49,43,93,75,57,29.86,29...        NaN       NaN   
1                                            2013-12-22       56.0      48.0   
2                                            2013-12-23       58.0      45.0   
3                                            2013-12-24       61.0      46.0   
4                                            2013-12-25       58.0      50.0   
...                                                 ...        ...       ...   
1314                                         2017-07-27      103.0      89.0   
1315                                         2017-07-28      105.0      91.0   
1316                                         2017-07-29      107.0      92.0   
1317                                         2017-07-30      106.0      93.0   
1318                                         2017-07-31       99.0      88.0   

      TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
0          NaN            NaN           NaN           NaN   
1         39.0           43.0          36.0          28.0   
2         32.0           31.0          27.0          23.0   
3         31.0           36.0          28.0          21.0   
4         41.0           44.0          40.0          36.0   
...        ...            ...           ...           ...   
1314      75.0           71.0          67.0          61.0   
1315      76.0           71.0          64.0          55.0   
1316      77.0           72.0          64.0          55.0   
1317      79.0           70.0          68.0          63.0   
1318      77.0           66.0          61.0          54.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  ...  \
0                     NaN                 NaN                 NaN  ...   
1                    93.0                68.0                43.0  ...   
2                    76.0                52.0                27.0  ...   
3                    89.0                56.0                22.0  ...   
4                    86.0                71.0                56.0  ...   
...                   ...                 ...                 ...  ...   
1314                 82.0                54.0                25.0  ...   
1315                 87.0                54.0                20.0  ...   
1316                 82.0                51.0                19.0  ...   
1317                 69.0                48.0                27.0  ...   
1318                 64.0                43.0                22.0  ...   

      VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0                    NaN                 NaN          NaN         NaN   
1                   10.0                 5.0         16.0         6.0   
2                   10.0                10.0          8.0         3.0   
3                   10.0                 7.0         12.0         4.0   
4                   10.0                 7.0         10.0         2.0   
...                  ...                 ...          ...         ...   
1314                10.0                10.0         12.0         5.0   
1315                10.0                10.0         14.0         5.0   
1316                10.0                10.0         12.0         4.0   
1317                10.0                10.0         13.0         4.0   
1318                10.0                10.0         12.0         4.0   

      WindGustMPH  Events  Snow  Fog  Rain Thunderstorm  
0             NaN     NaN     0    0     0            0  
1            25.0             0    0     0            0  
2            12.0             0    0     0            0  
3            20.0             0    0     0            0  
4            16.0             0    0     0            0  
...           ...     ...   ...  ...   ...          ...  
1314         21.0             0    0     0            0  
1315         20.0             0    0     0            0  
1316         17.0             0    0     0            0  
1317       

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [32]:
# Your code here
for i in event_list:
    austin_fixed[i]=austin_fixed['Events'].str.contains(i).astype(int)

ValueError: cannot convert float NaN to integer

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [ ]:
# Your code here


#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [ ]:
# Your code here


# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [ ]:
# Your code here


#### Print `austin_fixed` to check your `Date` column.

In [ ]:
austin_fixed.head(5)

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [ ]:
# Your code here:


In the next cell, import `train_test_split` from `sklearn.model_selection`

In [ ]:
#Your code here:


Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [ ]:
#Your code here:


#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:
